### PR TITLE
refactor: `BlockHeader`

### PR DESCRIPTION
**Pull request by aatifsyed**
Refactor `Blockheader` into a semantically equivalent `RawBlockHeader` (plain old data) and `CachingBlockHeader`, without the `#[derive(Builder)]` middleman and all those accessors. The former struct is now POD. 

Other changes
- replace the manual ser/de implementation with derives
- propagate `Hash` etc
- rename some methods
- fixup the module organisation

### CODE REVIEW

File: src/shim/sector.rs [hunk 1]
```diff
@@ -267,10 +268,24 @@ impl From<SectorSize> for SectorSizeV3 {
...
     derive_more::Into,
+    Eq,
 )]
 #[cfg_attr(test, derive(derive_quickcheck_arbitrary::Arbitrary))]
 pub struct PoStProof(PoStProofV3);
 
+impl Hash for PoStProof {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        // TODO(aatifsyed): missing from fvm_shared, as well as somewhat
+        //                  less-crucial (for us) PartialOrd/Ord implementations
```
Comment (aatifsyed) [inline src/shim/sector.rs:279]: fvm_shared::PoStProof should implement `Hash`, `Ord` and `PartialOrd`

File: src/blocks/header.rs [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+            }))
+        } else {
+            Ok(None)
+        }
+    }
+    pub fn cid(&self) -> &Cid {
+        self.cid.get_or_init(|| self.uncached.cid())
+    }
+
+    pub fn verify_signature_against(&self, addr: &Address) -> Result<(), Error> {
```
Comment (aatifsyed) [inline src/blocks/header.rs:225]: I don't understand why this is OK, but if it isn't that belongs in a separate PR

File: src/blocks/header.rs [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+                        .store(true, Ordering::Release);
+                    Ok(())
+                }
+                Err(e) => Err(e),
+            },
+        }
+    }
+}
+
+impl Serialize for CachingBlockHeader {
```
Comment (hanabi1224) [inline src/blocks/header.rs:243]: How about 
```rust
#[derive(Debug, Default, Serialize, Deserialize)]
pub struct CachingBlockHeader {
    #[serde(flatten)]
    uncached: RawBlockHeader,
    #[serde(skip)]
    cid: OnceCell<Cid>,
    #[serde(skip)]
    has_ever_been_verified_against_any_signature: AtomicBool,
}
```

**Review by hanabi1224 — APPROVED** [review]

**Review by LesnyRumcajs — APPROVED** [review]

**Review by LesnyRumcajs — APPROVED** [review]

**Review by hanabi1224 — APPROVED** [review]
